### PR TITLE
Wire dart2js through flutter tool, add compilation test

### PR DESCRIPTION
**Pull request by jonahwilliams**
Allows compilation to a stub dart:ui implementation for testing/tooling.

Requires https://github.com/flutter/engine/pull/7739 for missing members + libraries

### CODE REVIEW

File: packages/flutter/lib/src/foundation/internal.dart
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/internal.dart]: Is this ok? Wondering what the implications are for just taking your app and re-compiling it for the web?

File: packages/flutter/lib/src/foundation/internal_web.dart
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/internal_web.dart]: Did you ever figure out how dartdocs deals with conditional imports?

File: packages/flutter/lib/src/foundation/internal_web.dart
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/internal_web.dart]: Should this get the same doc comment as `internal.dart`? It strange that the comment is `The largest SMI value`, but the value is not that.

Could we use dartdoc templates here to make it very clear for people reading the code that this should be documented with the same comment?

File: packages/flutter/lib/src/foundation/basic_types.dart
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/basic_types.dart]: nit: I believe there should be a blank line separating import types.

File: packages/flutter/lib/src/foundation/basic_types.dart
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/basic_types.dart]: `internal` feel like an odd name choice here - especially for something that's exported right away?

File: packages/flutter_tools/lib/src/commands/build_web.dart
Comment (goderbauer) [inline packages/flutter_tools/lib/src/commands/build_web.dart]: Maybe indicate that this is experimental or something (I assume that's the reason it's hidden above)?

File: packages/flutter_tools/lib/src/web/compile.dart
Comment (goderbauer) [inline packages/flutter_tools/lib/src/web/compile.dart]: Can you make check that this will also work on Windows?

File: packages/flutter/lib/src/foundation/internal.dart
Comment (jonahwilliams) [reply packages/flutter/lib/src/foundation/internal.dart]: Long term we should deprecate and remove this class, but generally I think changing behavior is inescapable (FileImage, et cetera). The alternative is two separate programs.

File: packages/flutter_tools/lib/src/web/compile.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/web/compile.dart]: I do not have a windows machine to verify with

File: packages/flutter/lib/src/foundation/internal_web.dart
Comment (jonahwilliams) [reply packages/flutter/lib/src/foundation/internal_web.dart]: Currently it does not include the (dart.html) import, which is what we want. Should we put any doc comments on the unsupported implementation?

File: packages/flutter/lib/src/foundation/bitfield_unsupported.dart
Comment (jonahwilliams) [inline packages/flutter/lib/src/foundation/bitfield_unsupported.dart]: These do not show up in dartdocs.

File: packages/flutter_tools/lib/src/commands/build_web.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/build_web.dart]: Add all caps to make it scary

File: packages/flutter_tools/lib/src/commands/build_web.dart
Comment (yjbanov) [inline packages/flutter_tools/lib/src/commands/build_web.dart]: nit: "a Web"

**Review by yjbanov — APPROVED** [review]
LGTM from the Flutter Web perspective, but I'll let @goderbauer review from the tools perspective.

File: packages/flutter/lib/src/foundation/basic_types.dart
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/basic_types.dart]: should this be bitfield_unsupported.dart?

Comment (jonahwilliams) [reply packages/flutter/lib/src/foundation/basic_types.dart]: ....why did this pass the compile test

Comment (goderbauer) [reply packages/flutter/lib/src/foundation/basic_types.dart]: I'd be curious to know as well :)

**Comment by jonahwilliams** [comment]
We've determined that conditional exports are not handled correctly.

File: packages/flutter/lib/src/foundation/basic_types.dart
Comment (yjbanov) [reply packages/flutter/lib/src/foundation/basic_types.dart]: Wow! In that case, I recommend that you `flutter run --release` before merging the PR, in case Dart native AOT is also not handling conditional exports properly.

**Comment by yjbanov** [comment]
> We've determined that conditional exports are not handled correctly.

Is it something @vsmenon can help with?

File: packages/flutter/lib/src/foundation/basic_types.dart
Comment (jonahwilliams) [reply packages/flutter/lib/src/foundation/basic_types.dart]: Updating this to the correct import/export did not work as expected

**Comment by jonahwilliams** [comment]
The test did fail correctly but I also did not set the correct exitCode in the tool

**Comment by jonahwilliams** [comment]
Should I remove the config imports, and roll this with the test expected to fail?

**Comment by goderbauer** [comment]
What would be the value of that?

**Comment by jonahwilliams** [comment]
We'd at least have a way to keep track of what is currently blocking us

**Comment by vsmenon** [comment]
There is an open bug with conditional exports in particular:

https://github.com/dart-lang/sdk/issues/34179

Are imports working?

**Comment by yjbanov** [comment]
> Should I remove the config imports, and roll this with the test expected to fail?

I'm a fan of tests that are expected to fail! They are good at reminding you to later add a test that expects it to not fail.

**Comment by jonahwilliams** [comment]
I've tried a few different combinations of importing/exporting and I cant reproduce the successful compile anymore. It seems as if the imports are ignore by dart2js, or we're not using the config imports correctly.

**Comment by jonahwilliams** [comment]
This might just be a case of dart2js being too eager with detecting overflowing integer values

File: dev/bots/test.dart
Comment (goderbauer) [inline dev/bots/test.dart]: This should probably get a TODO to indicate that nonZero is in-fact not the correct behavior, but currently expected due to bugs (ideally linking to an issue that needs to get resolved to unskip this).

**Review by goderbauer — COMMENTED** [review]
I am not really convinced that checking this in in a broken state has value. Do we have a list of things that need to get resolved to actually make this useful?

**Comment by jonahwilliams** [comment]
Holding off on this til I have a bug to link to against dart2js or config imports

**Comment by jonahwilliams** [comment]
This was a big case of holding it wrong, everything is working now

**Pull request merged**
